### PR TITLE
Change flytekit Pytorch, TFJob and MPI plugins to use new kubeflow config

### PR DESCRIPTION
**Pull request by yubofredwang**
# TL;DR
We have updated the plugin config in flyteidl and flytepropeller. This PR updates flytekit to uses the new configs for kubeflow plugins including: pytorch, tfjob and mpijob.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
This change is backward compatible. If users continue using the old config like
```
@task(
task_config=TFJob(num_workers=1)
)
def my_task:
      ...
```
This still works.

However, the new config should look like:
```
@task(
task_config=TFJob(worker=Worker(replicas=1))
)
def my_task:
      ...
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3308

### CODE REVIEW

File: plugins/flytekit-kf-tensorflow/flytekitplugins/kftensorflow/models/chief.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    def restart_policy(self) -> Optional[RestartPolicy]:
+        return self._restart_policy
+
+    def to_flyte_idl(self) -> tensorflow_task.DistributedTensorflowTrainingReplicaSpec:
+        return tensorflow_task.DistributedTensorflowTrainingReplicaSpec(
+            replicas=self.replicas,
+            image=self.image,
+            resources=self.resources.to_flyte_idl() if self.resources else None,
+            restart_policy=self.restart_policy.value if self.restart_policy else None,
+        )
```
Comment (pingsutw) [inline plugins/flytekit-kf-tensorflow/flytekitplugins/kftensorflow/models/chief.py:58]: @yubofredwang Thanks for working on this
We plan to remove all the model files in flytekit, could you just use it inline [here](https://github.com/flyteorg/flytekit/pull/1627/files#diff-975a133e6c3847128c36fabd5b98f85aded0bed060412643ef44624707e179a6R117)

Comment (yubofredwang) [reply plugins/flytekit-kf-tensorflow/flytekitplugins/kftensorflow/models/chief.py:58]: Do you still want to keep the model class or it should be moved to inner logic to get_custom?

Comment (pingsutw) [reply plugins/flytekit-kf-tensorflow/flytekitplugins/kftensorflow/models/chief.py:58]: we don't need model class anymore, just move to inner logic to get_custom

Comment (yubofredwang) [reply plugins/flytekit-kf-tensorflow/flytekitplugins/kftensorflow/models/chief.py]: Updated, please verify if this is something you want.

File: plugins/flytekit-kf-tensorflow/README.md [hunk 2]
```diff
@@ -8,4 +8,16 @@ To install the plugin, run the following command:
...
+To migrate from v0 to v1, change the following:
+1. Update flytepropeller to v
+2. Update flytekit version to v
+3. Update your code from:
+    ```
+    task_config=TfJob(num_workers=10, num_ps_replicas=1, num_chief_replicas=1),
+    ```
+    to:
+    ```
+    task_config=TfJob(worker=Worker(replicas=10), ps=PS(replicas=1), chief=Chief(replicas=1)),
```
Comment (pingsutw) [inline plugins/flytekit-kf-tensorflow/README.md:22]: Could you add `RunPolicy` in this example?

File: plugins/flytekit-kf-tensorflow/README.md [hunk 3]
```diff
@@ -8,4 +8,16 @@ To install the plugin, run the following command:
 pip install flytekitplugins-kftensorflow
 ```
 
-_Example coming soon!_
+## Upgrade TensorFlow Plugin
+Tensorflow plugin is now updated from v0 to v1 to enable more configuration options.
+To migrate from v0 to v1, change the following:
+1. Update flytepropeller to v
```
Comment (pingsutw) [inline plugins/flytekit-kf-tensorflow/README.md:14]: v1.6.0. we just cut a release. https://github.com/flyteorg/flyte/blob/master/CHANGELOG/CHANGELOG-v1.6.0.md

File: plugins/flytekit-kf-tensorflow/README.md [hunk 4]
```diff
@@ -8,4 +8,16 @@ To install the plugin, run the following command:
...
+2. Update flytekit version to v
```
Comment (pingsutw) [inline plugins/flytekit-kf-tensorflow/README.md:15]: v1.6.1. next patch release.

**Review by pingsutw — COMMENTED** [review]
Looks great, thank you. some minor comments below.

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 5]
```diff
@@ -72,11 +138,42 @@ def __init__(self, task_config: PyTorch, task_function: Callable, **kwargs):
...
+            backoff_limit=run_policy.active_deadline_seconds,
+        )
+        
     def get_custom(self, settings: SerializationSettings) -> Dict[str, Any]:
-        job = DistributedPyTorchTrainingTask(workers=self.task_config.num_workers)
-        return MessageToDict(job)
+        worker = self._convert_replica_spec(self.task_config.worker)
+        # support v0 config for backwards compatibility
+        if (self.task_config.num_workers):
+            worker.replicas = self.task_config.num_workers
```
Comment (hamersaw) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:168]: Only comment is this (and similar in other plugins) means that the `v0` configuration overrides the `v1` configuration right? So
```
task_config=TfJob(
    num_workers=4,
    workers=WorkerSpec{
        replicas=3,
    },
)
```
would result in the number of workers being 4 right? Is this correct? I'm very happy to defer this decision.

Also, very much butchered the syntax, but I'm sure the idea is understood.

Comment (pingsutw) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:168]: maybe we can raise an error if users pass both v0 and v1 config.

Comment (yubofredwang) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:168]: v0 configuration overrides the v1 configuration. that's correct, so in that case it will be 4 workers. Do we want a different overriding logic?

**Review by pingsutw — COMMENTED** [review]
Overall LGTM, thanks

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 5 above)
Comment (eapolinario) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:168]: > maybe we can raise an error if users pass both v0 and v1 config.

I second that. We should simply disallow setting both configs.

File: plugins/flytekit-kf-tensorflow/README.md [hunk 6]
```diff
@@ -8,4 +8,47 @@ To install the plugin, run the following command:
...
+def tf_job():
+    ...
+```
+
+
+## Upgrade TensorFlow Plugin from  to V1
+Tensorflow plugin is now updated from v0 to v1 to enable more configuration options.
+To migrate from v0 to v1, change the following:
+1. Update flytepropeller to v1.6.0
+2. Update flytekit version to v1.6.1
```
[the remark this answers is not in the document]
Comment (eapolinario) [reply plugins/flytekit-kf-tensorflow/README.md:46]: 1.6.1 is the patch version released yesterday: https://pypi.org/project/flytekit/#history

File: plugins/flytekit-kf-tensorflow/setup.py [hunk 7]
```diff
@@ -5,7 +5,7 @@
 microlib_name = f"flytekitplugins-{PLUGIN_NAME}"
 
 # TODO: Requirements are missing, add them back in later.
-plugin_requires = ["flytekit>=1.3.0b2,<2.0.0"]
+plugin_requires = ["flytekit>=1.6.1", "flyteidl>=1.5.5"]
```
Comment (eapolinario) [inline plugins/flytekit-kf-tensorflow/setup.py:8]: any reason for specifying flyteidl here?

**Review by eapolinario — COMMENTED** [review]
sorry for the extra bit of process, but can you update the `requirements.txt` file in each of the plugins? This is going to get rid of the plugin test errors in CI. Make sure to use python 3.9.

File: plugins/flytekit-kf-tensorflow/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (yubofredwang) [reply plugins/flytekit-kf-tensorflow/README.md]: Updated

File: plugins/flytekit-kf-tensorflow/setup.py (hunk 7 above)
Comment (yubofredwang) [reply plugins/flytekit-kf-tensorflow/setup.py]: removed. makes sense to remove them

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 5 above)
Comment (yubofredwang) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:167]: added throwing error if users pass both v0 and v1 config.

**Comment by yubofredwang** [comment]
> sorry for the extra bit of process, but can you update the `requirements.txt` file in each of the plugins? This is going to get rid of the plugin test errors in CI. Make sure to use python 3.9.

Updated to use python 3.9. I noticed all other plugins are using python 3.8.

File: plugins/flytekit-kf-mpi/flytekitplugins/kfmpi/task.py [hunk 8]
```diff
@@ -110,34 +140,110 @@ class MPIFunctionTask(PythonFunctionTask[MPIJob]):
...
+
+    def _convert_run_policy(self, run_policy: RunPolicy) -> kubeflow_common.RunPolicy:
+        return kubeflow_common.RunPolicy(
+            clean_pod_policy=run_policy.clean_pod_policy.value if run_policy.clean_pod_policy else None,
+            ttl_seconds_after_finished=run_policy.ttl_seconds_after_finished,
+            active_deadline_seconds=run_policy.active_deadline_seconds,
+            backoff_limit=run_policy.active_deadline_seconds,
+        )
+
+    def _get_base_command(self, settings: SerializationSettings) -> List[str]:
```
Comment (eapolinario) [inline plugins/flytekit-kf-mpi/flytekitplugins/kfmpi/task.py:186]: just curious, why do we need this method?

**Review by eapolinario — APPROVED** [review]
one minor question, otherwise LGTM.

File: plugins/flytekit-kf-mpi/flytekitplugins/kfmpi/task.py (hunk 8 above)
Comment (yubofredwang) [reply plugins/flytekit-kf-mpi/flytekitplugins/kfmpi/task.py:186]: Good question. So HorovodFunctionTask extends MPIFunctionTask extends PythonFunctionTask, and get_command get overriden by MPIFunctionTask.

In HorovodFunctionTask, I would like to use the get_command() from PythonFunctionTask, thus I have to create yet another function in MPIFunctionTask to do so.

**Comment by eapolinario** [comment]
> > sorry for the extra bit of process, but can you update the `requirements.txt` file in each of the plugins? This is going to get rid of the plugin test errors in CI. Make sure to use python 3.9.
> 
> Updated to use python 3.9. I noticed all other plugins are using python 3.8. build-plugins 3.8 tests will fail in this case

I apologize for flip-flopping on this. I checked the requirements files for the 3 plugins and couldn't see why it wouldn't work for python 3.8, so I took the liberty of regenerating the requirements files. Will merge after tests pass.

**Review by eapolinario — APPROVED** [review]

**Review by hamersaw — APPROVED** [review]
